src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs

```diff
@@ -0,0 +1,178 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    // https://tools.ietf.org/html/rfc1928
+    internal static class SocksHelper
+    {
+        // socks protocol limits address length to 1 byte, thus the maximum possible buffer is 256+other fields
+        private const int BufferSize = 512;
+        private const int ProtocolVersion = 5;
+        private const byte METHOD_NO_AUTH = 0;
+        // private const byte METHOD_GSSAPI = 1;
+        private const byte METHOD_USERNAME_PASSWORD = 2;
+        private const byte METHOD_NO_ACCEPTABLE = 0xFF;
+        private const byte CMD_CONNECT = 1;
+        // private const byte CMD_BIND = 2;
+        // private const byte CMD_UDP_ASSOCIATE = 3;
+        private const byte ATYP_IPV4 = 1;
+        private const byte ATYP_DOMAIN_NAME = 3;
+        private const byte ATYP_IPV6 = 4;
+        private const byte REP_SUCCESS = 0;
+        // private const byte REP_FAILURE = 1;
+        // private const byte REP_NOT_ALLOWED = 2;
+        // private const byte REP_NETWORK_UNREACHABLE = 3;
+        // private const byte REP_HOST_UNREACHABLE = 4;
+        // private const byte REP_CONNECTION_REFUSED = 5;
+        // private const byte REP_TTL_EXPIRED = 6;
+        // private const byte REP_CMD_NOT_SUPPORT = 7;
+        // private const byte REP_ATYP_NOT_SUPPORT = 8;
+
+        public static async ValueTask EstablishSocks5TunnelAsync(Stream stream, string host, int port, Uri proxyUri, ICredentials? proxyCredentials, CancellationToken cancellationToken)
+        {
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(BufferSize);
+
+            try
+            {
+                // +----+----------+----------+
+                // |VER | NMETHODS | METHODS  |
+                // +----+----------+----------+
+                // | 1  |    1     | 1 to 255 |
+                // +----+----------+----------+
+                buffer[0] = ProtocolVersion;
+                var credentials = proxyCredentials?.GetCredential(proxyUri, "");
```

Comment:
`ICredentials` was designed for HTTP authentication. The authType parameter is not defined for others.

Reply:
We can pass `null` here since the parameter is nullable

Reply:
It's not nullable